File: src/main/java/ee/tenman/portfolio/alphavantage/SimplifiedRateLimiter.java
package ee.tenman.portfolio.alphavantage;

import com.google.common.util.concurrent.RateLimiter;

public class SimplifiedRateLimiter {
	private static final double RATE = 29.0 / 60.0; // 29 requests per minute
	private final RateLimiter rateLimiter;
	
	public SimplifiedRateLimiter() {
		this.rateLimiter = RateLimiter.create(RATE);
	}
	
	public void execute(Runnable task) {
		rateLimiter.acquire(); // This will block if necessary to ensure the rate
		task.run();
	}
}

File: src/main/kotlin/ee/tenman/portfolio/alphavantage/SimplifiedRateLimiter.kt
package ee.tenman.portfolio.alphavantage

import com.google.common.util.concurrent.RateLimiter

class SimplifiedRateLimiter {
  companion object {
    private const val RATE = 29.0 / 60.0 // 29 requests per minute
  }

  private val rateLimiter: RateLimiter = RateLimiter.create(RATE)

  fun execute(task: Runnable) {
    rateLimiter.acquire() // This will block if necessary to ensure the rate
    task.run()
  }
}
